Add OAuth 2.0 authentication support

zparnold opened the pull request:
<!-- Thank you for your contribution! Please provide a brief summary. -->

## Description

<!-- What does this PR do? Why is it needed? -->
<!-- Link related issues: Fixes #<issue_number> -->

This PR adds OAuth 2.0 authentication support for Atlassian Cloud services (Jira and Confluence). This implementation allows users to authenticate using the modern OAuth 2.0 flow with Atlassian Cloud, which is more secure and flexible than API tokens, especially for integrations with third-party applications.

Fixes: #238

## Changes

<!-- Briefly list the key changes made. -->

- Added OAuth 2.0 configuration and token management in `utils/oauth.py`
- Implemented an OAuth authorization helper script (`scripts/oauth_authorize.py`)
- Updated Jira and Confluence clients to support OAuth authentication
- Added environment variable support for OAuth configuration
- Updated documentation with OAuth setup instructions
- Added comprehensive test coverage for OAuth functionality

## Testing

<!-- How did you test these changes? (e.g., unit tests, integration tests, manual checks) -->

- [x] Unit tests added/updated
  - Added test_client_oauth.py for both Jira and Confluence
  - Added test_oauth.py for OAuth utility functions
- [ ] Manual checks performed: `[Tested full OAuth flow with Atlassian Cloud, verified token refresh, and API access]` Will complete this once I have a tenant to test with

## Checklist

- [x] Code follows project style guidelines (linting passes).
- [x] Tests added/updated for changes.
- [x] All tests pass locally.
- [x] Documentation updated (if needed).

sooperset commented on src/mcp_atlassian/utils/oauth.py:
```diff
@@ -0,0 +1,307 @@
+            token_dir = Path.home() / ".mcp-atlassian"
+            token_dir.mkdir(exist_ok=True)
+
+            # Save the tokens to a file
+            token_path = token_dir / f"oauth-{self.client_id}.json"
```
For storing refresh tokens, you might consider using the `keyring` library instead of JSON files. This would leverage the operating system's secure credential storage and might be a more robust approach for handling sensitive tokens.

zparnold replied:
great idea! thanks!

sooperset commented on src/mcp_atlassian/utils/oauth.py:
```diff
@@ -0,0 +1,307 @@
+        """
+        try:
+            # Create the directory if it doesn't exist
+            token_dir = Path.home() / ".mcp-atlassian"
```
If you decide to use keyring for storage, the token loading function would need to be updated as well to retrieve from the secure storage instead of the JSON file.

sooperset commented on scripts/oauth_authorize.py:
```diff
@@ -0,0 +1,308 @@
+#!/usr/bin/env python
+"""
+OAuth 2.0 Authorization Flow Helper for MCP Atlassian
```
This helper script works well for handling user interaction. To improve user experience, you might consider adding an entry point like `mcp-atlassian --oauth-setup`. Then users don't need to clone this repo and run the script.

zparnold replied:
good suggestion will do

sooperset reviewed (commented):
@zparnold Thank you for this PR adding OAuth 2.0 authentication support! This addresses issue #238 and provides a more secure and flexible authentication option for Atlassian Cloud users.

I'm impressed with your implementation structure. You've done an excellent job separating the complex OAuth flow into a dedicated helper script and encapsulating core OAuth logic in the appropriate modules.

I have a suggestion regarding token storage that might enhance security. Currently, the implementation stores tokens in JSON files, but have you considered using the keyring library? It would store credentials in the OS's native secure storage (like macOS Keychain or Windows Credential Manager) which could provide better protection for sensitive tokens.

zparnold commented:
Thank you for your valuable feedback! I've implemented the suggested changes:

1. **Secure Token Storage with Keyring**: 
   - Added keyring integration to store OAuth tokens in the operating system's native secure credential storage (macOS Keychain, Windows Credential Manager, or Linux Secret Service)
   - Implemented fallback to file-based storage if keyring is not available
   - All related tests are now passing

2. **Improved User Experience with Built-in OAuth Setup Wizard**:
   - Added a convenient `--oauth-setup` command-line flag to the main CLI
   - Created an interactive setup wizard that guides users through the OAuth configuration process
   - Users can now easily set up OAuth authentication by running `mcp-atlassian --oauth-setup`

3. **Fixed CI Test Issue**:
   - Updated the test_init_with_basic_auth test to correctly expect the verify_ssl parameter

The implementation provides a secure and user-friendly way for users to authenticate with Atlassian Cloud services using OAuth 2.0. The tokens are now stored securely by default, with a fallback mechanism for environments where keyring is not available.

sooperset reviewed (approved):
LGTM! Thanks for implementing these changes and for your great contribution!
